src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/CustomRecordTypeWithUndefinedFields.cs

```diff
@@ -0,0 +1,200 @@
+
+            _customRecordType = new TreatMissingFieldAsUntypedRecordType(_originalRecordType);
+            _testObj = new TestObj
+            {
+                Properties = new Dictionary<string, object>
+                {
+                    { "prop1",  1 },
+                    { "prop2",  "hello" },
+                    { "prop_not_defined_in_type",  3 },
+                }
+            };
+        }
+
+        [Theory]
+        [InlineData("obj.prop1", 1.0, false)]
+        [InlineData("IsError(obj.prop2)", true, false)] // type mismatch should fail at runtime
+        [InlineData("IsBlank(obj.missing)", null, true)] // type checking fails when referencing undefined type.
+        [InlineData("IsBlank(obj.missing.missing)", null, true)]
+        [InlineData("IsBlank(obj.prop_not_defined_in_type)", null, true)]
+        [InlineData("IsError(obj.missing + 1)", null, true)]
+        public async Task TestOriginalRecordType(string expr, object expected, bool hasException)
+        {
+            var engine = new RecalcEngine();
+            var symbolTable = new SymbolTable();
+            var symbolValues = new SymbolValues();
+
+            symbolTable.AddVariable("obj", _originalRecordType);
+            symbolValues.Add("obj", new TestCustomRecordValue(_testObj, _originalRecordType));
+
+            try
+            {
+                var result = await engine.EvalAsync(expr, default, null, symbolTable, symbolValues).ConfigureAwait(false);
+                Assert.False(hasException);
+                Assert.Equal(expected, result.ToObject());
+            }
+            catch (InvalidOperationException)
+            {
+                Assert.True(hasException);
+            }
+        }
+
+        [Theory]
+        [InlineData("obj.prop1", 1.0, false)]
+        [InlineData("IsError(obj.prop2)", true, false)] // type mismatch should fail at runtime.
+        [InlineData("IsBlank(obj.missing)", true, false)] // fields not defined in RecordType is treated as Blank.
+        [InlineData("IsBlank(obj.missing.missing)", true, false)]
+        [InlineData("IsBlank(obj.prop_not_defined_in_type)", true, false)] // fields not defined in RecordType is always Blank, regardless of whether it is defined in actual data.
+        [InlineData("IsError(obj.missing + 1)", null, true)] // type checking fails when operating on undefined field.
+        public async Task TestCustomRecordType(string expr, object expected, bool hasException)
+        {
+            var engine = new RecalcEngine();
+            var symbolTable = new SymbolTable();
+            var symbolValues = new SymbolValues();
+
+            symbolTable.AddVariable("obj", _customRecordType);
+            symbolValues.Add("obj", new TestCustomRecordValue(_testObj, _customRecordType));
+
+            try
+            {
+                var result = await engine.EvalAsync(expr, default, null, symbolTable, symbolValues).ConfigureAwait(false);
+                Assert.False(hasException);
```

Comment:
> Assert.False(hasException); [](http://example.com/codeflow?start=15&length=28)

Note we could actually split the eval up into parts to be more specific and avoid the exception:

CheckResult check = engine.Check(expr, ...) 
Assert.Equals(check.IsSuccess, !hasException);

if (check.IsSuccess) {
    // Now we eval it, 
   var run = eval.GetEvaluator();
   var result = run.Eval(symbolValues);
   Assert.Equal(expected, result.ToObject());
}


This could actually be important if you want to use the same compiled expression against multiple inputs.

Reply:
Thanks! Updated accordingly.